### PR TITLE
UX: Added leading whitespace to emoji picker & inline translation for…

### PR DESCRIPTION
**Pull request by venarius**
… emoji escape

This enhances the emoji picker to now automatically append a whitespace in front of the "about to insert" emoji, in order to get the emojis to escape correctly.

Also `emojiUnescape` and `emojiEscape` now support the `enable_inline_emoji_translation` site setting.

Resolves: https://meta.discourse.org/t/emoji-not-recognized-after-cyrillic-symbols/72091/17

### CODE REVIEW

**Comment by jjaffeux** [comment]
@venarius both tests failures seems legit and related to the commit

**Comment by venarius** [comment]
@jjaffeux yes, already working on it. That's why it's still WIP

**Comment by jjaffeux** [comment]
> @jjaffeux yes, already working on it. That's why it's still tagged as WIP

Yes sorry saw the WIP after ;) I'm tracking failures in PR lately as we have some tests failing randomly, making sure it's not blocking you.

**Comment by venarius** [comment]
> Also do you know about the draft PR feature ? https://github.blog/2019-02-14-introducing-draft-pull-requests/

Oh, nope, didn't know about them :) The only thing I saw people here in the pull requests doing was adding a "Do not merge" badge.

> Yes sorry saw the WIP after ;) I'm tracking failures in PR lately as we have some tests failing randomly, making sure it's not blocking you.

Haha all fine :p For me it's pretty normal that discourse tests are failing to be honest. When I started contributing, every second PR the frontend tests failed for me because of some "kit" not being "expanded" ?!

**Comment by jjaffeux** [comment]
> every second PR the frontend tests failed for me because of some "kit" not being "expanded" ?!

yes I did improve it weeks ago, and have a lot more work provisioned on this

File: app/assets/javascripts/pretty-text/emoji.js.es6 [hunk 1]
```diff
@@ -43,43 +43,72 @@ export function performEmojiUnescape(string, opts) {
     return;
   }
 
-  return string.replace(unicodeRegexp, m => {
-    const isEmoticon = !!translations[m];
-    const isUnicodeEmoticon = !!replacements[m];
+  const inlineEmojiEnabled =
+    typeof Discourse !== "undefined"
+      ? Discourse.SiteSettings.enable_inline_emoji_translation
```
Comment (ZogStriP) [inline app/assets/javascripts/pretty-text/emoji.js.es6]: Why is this not an option? Like `opts.inlineEmojiEnabled`?

File: app/assets/javascripts/pretty-text/emoji.js.es6 [hunk 2]
```diff
@@ -43,43 +43,72 @@ export function performEmojiUnescape(string, opts) {
...
       ? "emoji emoji-custom"
       : "emoji";
-
-    return url && (isEmoticon || hasEndingColon || isUnicodeEmoticon)
-      ? `<img src='${url}' ${
-          opts.skipTitle ? "" : `title='${emojiVal}'`
-        } alt='${emojiVal}' class='${classes}'>`
-      : m;
-  });
+    const isAllowed =
```
Comment (ZogStriP) [inline app/assets/javascripts/pretty-text/emoji.js.es6]: This is very hard to read/process/understand. Can it be split in several lines with either comments or (better) carefully named variables?

File: app/assets/javascripts/pretty-text/emoji.js.es6 [hunk 3]
```diff
@@ -43,43 +43,72 @@ export function performEmojiUnescape(string, opts) {
...
+      (!inlineEmojiEnabled &&
+        (/\s|[>.,\/#!$%^&*;:{}=\-_`~()]/.test(before) || m.index === 0)) ||
+      !!inlineEmojiEnabled;
+
+    const replacement =
```
Comment (ZogStriP) [inline app/assets/javascripts/pretty-text/emoji.js.es6]: Same here, that's a huge ternary 😲

File: lib/pretty_text.rb [hunk 4]
```diff
@@ -230,20 +230,22 @@ def self.unescape_emoji(title)
 
     set = SiteSetting.emoji_set.inspect
     custom = Emoji.custom.map { |e| [e.name, e.url] }.to_h.to_json
+    inline_emoji = SiteSetting.enable_inline_emoji_translation
     protect do
       v8.eval(<<~JS)
         __paths = #{paths_json};
-        __performEmojiUnescape(#{title.inspect}, { getURL: __getURL, emojiSet: #{set}, customEmoji: #{custom} });
+        __performEmojiUnescape(#{title.inspect}, { getURL: __getURL, emojiSet: #{set}, customEmoji: #{custom}, inlineEmojiEnabled: #{inline_emoji} });
```
Comment (ZogStriP) [inline lib/pretty_text.rb]: I think splitting this line is easier on the 👀 

```suggestion
        __performEmojiUnescape(#{title.inspect}, { 
          getURL: __getURL, 
          emojiSet: #{set}, 
          customEmoji: #{custom}, 
          inlineEmojiEnabled: #{inline_emoji} 
        });
```

**Comment by tgxworld** [comment]
@jjaffeux Is it OK if you take over this PR and get it merged since you know the emoji code better? :grin:

**Comment by ZogStriP** [comment]
Still on your list @jjaffeux?

**Comment by ZogStriP** [comment]
@gschlager can you take over this one?

**Pull request closed without merging**